pkg/sentry/socket/plugin/cgo/socket_unsafe.go

```diff
@@ -0,0 +1,447 @@
+		return -linuxerrno.EINVAL
+	}
+	if ret := int(C.plugin_getpeername(
+		C.int(handle),
+		(*C.struct_sockaddr)(unsafe.Pointer(&addr[0])),
+		(*C.socklen_t)(unsafe.Pointer(addrlen)),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret < 0 {
+		return -int(errno)
+	} else {
+		return ret
+	}
+}
+
+// Readiness works as a CGO wrapper for plugin_readiness.
+func Readiness(handle uint32, mask uint64) int {
+	return int(C.plugin_readiness(C.int(handle), C.int(mask)))
+}
+
+// Read works as a CGO wrapper for plugin_read.
+func Read(handle uint32, buf uintptr, count int) int64 {
+	var errno uint64
+	if ret := int64(C.plugin_read(
+		C.int(handle),
+		unsafe.Pointer(buf),
+		C.size_t(count),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret < 0 {
+		return -int64(errno)
+	} else {
+		return ret
+	}
+}
+
+// Readv works as a CGO wrapper for plugin_readv.
+func Readv(handle uint32, iovs []syscall.Iovec) int64 {
+	var errno uint64
+	if len(iovs) == 0 {
+		return 0
+	}
+	if ret := int64(C.plugin_readv(
+		C.int(handle),
+		(*C.struct_iovec)(unsafe.Pointer(&iovs[0])),
+		C.int(len(iovs)),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret < 0 {
+		return -int64(errno)
+	} else {
+		return ret
+	}
+}
+
+// Recvfrom works as a CGO wrapper for plugin_recvfrom.
+func Recvfrom(handle uint32, buf, addr []byte, flags int) (int64, int) {
+	var errno uint64
+	addrlen := len(addr)
+	if ret := int64(C.plugin_recvfrom(
+		C.int(handle),
+		GetPtr(buf),
+		C.size_t(len(buf)),
+		C.int(flags),
+		(*C.struct_sockaddr)(GetPtr(addr)),
+		(*C.socklen_t)(unsafe.Pointer(&addrlen)),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret < 0 {
+		return -int64(errno), addrlen
+	} else {
+		return ret, addrlen
+	}
+}
+
+// Recvmsg works as a CGO wrapper for plugin_recvmsg.
+func Recvmsg(handle uint32, iovs []syscall.Iovec, addr, control []byte, flags int) (int64, int, int, int) {
+	lenAddr := len(addr)
+	lenCtl := len(control)
+	sysflags := flags | syscall.MSG_DONTWAIT
+
+	if len(iovs) == 0 {
+		return 0, lenAddr, lenCtl, 0
+	}
+
+	var ptrAddr, ptrCtl *byte
+	if lenAddr > 0 {
+		ptrAddr = &addr[0]
+	}
+
+	if lenCtl > 0 {
+		ptrCtl = &control[0]
+	}
+
+	msg := syscall.Msghdr{
+		Iov:        &iovs[0],
+		Iovlen:     uint64(len(iovs)),
+		Name:       ptrAddr,
+		Namelen:    uint32(lenAddr),
+		Control:    ptrCtl,
+		Controllen: uint64(lenCtl),
+	}
+
+	var errno uint64
+	if ret := int64(C.plugin_recvmsg(
+		C.int(handle),
+		(*C.struct_msghdr)(unsafe.Pointer(&msg)),
+		C.int(sysflags),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret < 0 {
+		return -int64(errno), lenAddr, lenCtl, 0
+	} else {
+		return ret, int(msg.Namelen), int(msg.Controllen), int(msg.Flags)
+	}
+}
+
+// Write works as a CGO wrapper for plugin_write.
+func Write(handle uint32, buf uintptr, count int) int64 {
+	var errno uint64
+	if ret := int64(C.plugin_write(
+		C.int(handle),
+		unsafe.Pointer(buf),
+		C.size_t(count),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret < 0 {
+		return -int64(errno)
+	} else {
+		return ret
+	}
```

Comment:
Thanks for advice :) Yes, it will be much cleaner to create a wrapper. We have added a wrapper called convertRetVal and replace all these patterns with the wrapper.